compiler/qsc_data_structures/src/line_column.rs

```diff
@@ -0,0 +1,192 @@
+            }
+
+            if line > self.line || (line == self.line && column > self.column) {
+                // We moved past the target line+column
+                return u32(byte_offset);
+            }
+        }
+
+        // return eof if we move past the end of the string
+        u32(contents.len())
+    }
+}
+
+/// Same as a [`Span`], but represented with [`Position`]s instead of offsets.
+#[derive(Debug, PartialEq, Clone, Copy, Eq, Hash)]
+pub struct Range {
+    pub start: Position,
+    pub end: Position,
+}
+
+impl Range {
+    /// For a given string and a span expressed in utf-8 byte offsets,
+    /// returns a [`Range`] that corresponds to that span.
+    ///
+    /// Any offsets that go past the end of the string are mapped to the end of the string.
+    /// (e.g. for "hello", the span (0-10) would map to (0,0-0,5)).
+    ///
+    /// This function does not validate span range. If `span.hi > span.lo`, the
+    /// end offset is simply mapped to the end of the string.
+    /// (e.g. for "hello", the invalid span (3-0) would map to (0,3-0,5)).
+    ///
+    /// This function also does not validate whether the passed in offset
+    /// is a valid utf8 char boundary. If an invalid offset is passed in,
+    /// the next char boundary will be returned.
+    #[must_use]
+    pub fn from_span(encoding: Encoding, contents: &str, span: &Span) -> Self {
+        let [start, end] = positions_from_utf8_byte_offsets(encoding, contents, [span.lo, span.hi]);
+        Self { start, end }
+    }
+
+    #[must_use]
+    pub fn empty(&self) -> bool {
+        self.start == self.end
+    }
+}
+
+/// For a given string and array of utf-8 byte offsets, returns the [`Position`]s
+/// corresponding to the byte offsets.
+///
+/// Since this function iterates over the string once, the array *MUST* be in sorted order.
+///
+/// This function takes a (const sized) array to avoid allocations.
+fn positions_from_utf8_byte_offsets<const N: usize>(
+    encoding: Encoding,
+    contents: &str,
+    sorted_utf8_byte_offsets: [u32; N],
+) -> [Position; N] {
+    // The below example contains characters that are encoded
+    // with more utf-8 code units (bytes) than utf-16 code units, to demonstrate
+    // how code unit offset will differ depending on the chosen encoding.
+
+    // chars                    | 𝑓                 (        𝑥                 ⃗        )     <eof>
+    // unicode code point       | 1d453             28       1d465             20d7     29
+    // utf-8 units (bytes)      | f09d9193          28       f09d91a5          e28397   29
+    // utf-16 units             | d835     dc53     28       d835     dc65     20d7     29
+    // char offset              | 0                 1        2                 3        4
+    // utf-8 byte offset        | 0                 4        5                 9        12
+    // utf-16 code unit offset  | 0                 2        3                 5        6
```

Comment:
UTF-16, as the name says, always has 16 bit code units, so there is never an odd offset. Looks like 3 & 5 are wrong here, and in line 134 above the code units should be 0x0028 and 0x0029, not 28 & 29.

Reply:
Ah! I guess those are 'unit' offsets, not 'byte' offsets, so my mistake. Still, might be clearer to show utf-16 code units as 0x0028, rather than 28, to highlight it is 2 bytes long.